### ISSUE TITLE
Annual report includes training records

### ISSUE THREAD

**Issue by JimBacon** [BiologicalRecordsCentre/ABLE#632]
The report needs to respect the training flag and exclude such records.

**Comment by andrewvanbreda** [comment]
@JimBacon 
I am looking at https://github.com/BiologicalRecordsCentre/ABLE/issues/515 so I will do this one at same time.

**Comment by andrewvanbreda** [comment]
Hi @Gary-van-Breda,

As discussed on phone, the way the pages that interact with the summary_builder module need altering (I think this is primarily the Annual Report.

To recap, on the User Account page there is a "Training" checkbox.
Normally on Indicia sites selecting this checkbox means the following

- A warning is displayed telling the user they are in this mode. This message appears on most pages as the user navigates the site.
- When on, the user's samples and occurrences are both marked as training during data entry.
- When on, the user is only shown training data on grids and downloads (and vice versa when it is off).

Currently the warning works, and the data entry should be working.

So our problem is the data shown to the user.

There seem to be 4 different situations regarding this.

1. Elastic Search pages seem to automatically detect that Training flag and work correctly.
So I have not had to do any work on this area.

2. There are simple SQL driven pages/downloads where I just need to make the usual kind of changes to reports to make them training flag sensitive.

3. There are some SQL reports where I have not been able to locate their use on the system.
As discussed on the phone, I will not change these reports as there is a chance that changes to these reports might cause breakages
e.g. as discussed on the phone, the occurrence download "gis" report is like this because you said it is opened in special way from a footer.

4. There are pages driven by the Summariser Builder module.
This includes a lot of the Annual Report.
Currently my understanding of this is that the data for this pages come from special tables generated by this module which are a mixture of training and non-training data.

In terms of what we do about the above:

1. Don't need to do anything, the Elastic pages seems to work already

2. I will finished changing pages and will attach a list of pages/reports I have changed to this issue.

3. Again I will supply a list of SQL reports. This will be a list of reports I have NOT changed but I think are relevant because they refer to samples/occurrences.

4. You will need to think of a solution to this problem.
As discussed on phone there are two possibilities.

A.
Change the pages interacting the Summary Building module so that those pages are training flag sensitive.

B.
If the above isn't possible, an alternative a solution might be to changes it so pages like the Annual Report to always only showing non-training data (then warn the user about this behaviour).
I guess if this solution is the only way, further discussion by all parties will be required as it might limit the ability of the site to be used for Annual Report user training.

**Comment by andrewvanbreda** [comment]
Hi @Gary-van-Breda ,

Here are things I think you will need to change to be training flag sensitive. I can't change them as they rely on data from the summary_builder module
Annual Report Download Summary Grid button (couldn't work out where the data was coming from)
Annual Report Download Estimates Grid button (couldn't work out where the data was coming from)
Rest of Annual Report (apart from the SQL report download buttons listed below that I was able to change)
ukbms_year_plot_index (My year by year index plot)
ebms_year_plot_count (not sure where this one is used, but uses summary_occurrences)

Here are the reports that I think might be relevant I have not changed because I was not sure if changes would cause breakages and I could not see where they were being used on the website.
I would be interested to know if you think I should change these.
(noting I have ignore UKBMS named reports as I have not tested training on that site yet).
ebms_occurrence_download
ebms_occurrence_point_gis_download_user
ebms_occurrence_point_gis_download
ebms_sample_gis_download_user
ebms_sample_gis_download
ebms_sample_quick_view
ebms_samples_species_565
ebms_samples_species_645
ebms_samples_species_combined_for_user
ebms_scheme_users_sample_details
ebms_species_counts_by_sample_for_survey
ebms_timed_count_samples_gis_download
moth_samples
samples_summary_list_mapping_lores
samples_summary_list_mapping
samples_summary_list

FYI
Here are all the reports/pages I have changed to be training flag sensitive that weren't before.
Note some of these are coded but not live yet
(Not that a lot of the reports were in report_for_prebuilt_forms, for any reports like that I created a new version of report with same name in indicia-reports/project).

ebms_my_walks_list (My Walks List, My Moth Trap Samples, My Moth Trap Samples)
ebms_my_region_walks_list (My Region Walks List)
ebms_summary_download (Annual Report->Downloads->Summary Level download button)
ebms_section_download (Annual Report->Downloads->Section Level download button)
ebms_sample_download (Annual Report->Downloads->Samples download button)
ebms_occurrence_download_AS (Annual Report->Downloads->Occurrences download button)
ebms_visits_and_sites_summary_3 (My Site/Visit Summary)
All sites (ukbms/all_sites)
ebms_annual_summary_raw_data (Annual Report->Raw data tab and download)
moth_traps (All moth traps page)
ukbms_section_plot (My section plot)
ukbms_branch_walks_list (Scheme Walks)
ukbms_visits_and_sites_summary (Global Site/Visit Summary)

Here are some pages I do intend to change but am waiting for John to answer some questions about the reports.
species_and_occurrence_counts_2_lvl_smp (My Transect Sites)
species_and_occurrence_counts_1_lvl_smp (My moth traps)
ebms_branch_sites (My region sites, Scheme Sites)


FYI here are Eastic Search pages we don't need to change as seem to pick up training flag ok
My Samples
My Downloads
Summary by country & year
Verification
Downloads
Explore Online Data

One last note. The "Scheme users information" page is driven by a report I wrote.
As this report includes training data and non-training actually in separate columns, this report does not look at the user's Training setting, and does not need changing.

**Comment by andrewvanbreda** [comment]
The live site is now configured for all the following reports (same reports I listed above):
ebms_my_walks_list (live and tested)
ebms_my_region_walks_list (live and tested)
ebms_summary_download (live and tested)
ebms_section_download (live and tested)
ebms_sample_download (live and tested)
ebms_occurrence_download_AS (live and tested)
ukbms_branch_walks_list (live and tested)
ebms_visits_and_sites_summary_3 (live and tested)
ukbms_section_plot (live and tested)
moth_traps (tested on multidev, configured on live site, changes will go live next time someone pulls reports folder)
All sites (tested on multidev, configured on live site, changes will go live next time someone pulls reports folder)
ebms_annual_summary_raw_data (tested on multidev, configured on live site, changes will go live next time someone pulls reports folder)
ukbms_visits_and_sites_summary  (tested on multidev, configured on live site, changes will go live next time someone pulls reports folder)

Note to get ukbms_section_plot to work, I had to had some code to the PHP file
    	if (hostsite_get_user_field('training')) {
          $options['extraParams']['training'] = 'true';
        }

The similar been added to Year Index Plot for when that report is done.
    	if (hostsite_get_user_field('training')) {
          $options['countExtraParams']['training'] = 'true';
          $options['indexExtraParams']['training'] = 'true';
        }

I guess when UKBMS is checked for the training flag, the same changes will need to be made there.

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod

Is it possible to send me this file on the Live Warehouse please
 I think it needs changing to exclude training records, but I need to make sure it hasn't already been customised.
/modules/summary_builder/config/summary_builder.php

Cheers

@Gary-van-Breda 
My plan is to simply exclude training from the queries in that file.
Let me know if you think that is wrong approach, it is hard to test on my machine, but I think this is right thing to do.

**Comment by BirenRathod** [comment]
@andrewvanbreda Here it is. 

[summary_builder.zip](https://github.com/user-attachments/files/19213574/summary_builder.zip)

**Comment by andrewvanbreda** [comment]
Hi @BirenRathod ,

I have an adjusted summary_builder config file attached which should exclude training.

Unfortunately my ability to test this is minimal because I have no EBMS data on my machine.

I would like to see how it behaves before I hotfix it.

It is possible to rename the existing summary_builder config file  /modules/summary_builder/config/summary_builder.php

to be summary_builder_old.php then put the unzipped attached file in that same folder?

Then I can check its behaviour and also get Cristina's thoughts before I commit it, and it can be reversed easily if needed.

I think this is fairly low risk as the changes are not logically complex, and I did check the SQL does run in PGadmin without error (even if the rows are empty). The Annual Report is not great anyway if it includes training data.

Let me know if the above is not possible and I have to commit it.

Cheers

Andy

[summary_builder.php.zip](https://github.com/user-attachments/files/19362405/summary_builder.php.zip)

**Comment by BirenRathod** [comment]
@andrewvanbreda  I can do that but before doing anything, you can check with @johnvanbreda first if any other impacts it is having.

**Comment by andrewvanbreda** [comment]
OK no problem. I am not sure if Summary Builder is used outside EBMS/UKBMS maybe John can confirm.
Note the summary_builder is not sensitive to the user's training mode setting, it simply makes the assumption that training data is never useful in the summary.

**Comment by johnvanbreda** [comment]
I don't use the summary builder outside those 2 projects.

**Comment by Gary-van-Breda** [comment]
You can see which websites are registered to used the summary builder on https://warehouse1.indicia.org.uk/index.php/summariser_definition

**Comment by BirenRathod** [comment]
thanks @Gary-van-Breda. That is helpful.
@johnvanbreda  I'm not aware of Surrey BIC & ERCCIS. is there implications of updating the sumarry_builder.php on those two sites?

**Comment by andrewvanbreda** [comment]
I think it is probably worth noting that the existing summary builder data includes training data, then that isn't good, so even if there was an issue with what I have done we are probably not in a worse position.

**Comment by BirenRathod** [comment]
@andrewvanbreda Considering the minimal risk, I will update the file unless anybody finds out not to do.

**Comment by andrewvanbreda** [comment]
Ok that is great.
I am not actually quite sure how often the summary_occurrences table get regenerated in totality, whether it is each scheduled_tasks run.
@Gary-van-Breda might be able to let me know (i.e. at what point are we going to see the result of the new code in the actual annual report itself).

**Comment by johnvanbreda** [comment]
As it's low risk I'm sure it's OK to update with respect to ERCCIS and Surrey BIC.

**Comment by andrewvanbreda** [comment]
@BirenRathod @Gary-van-Breda 
Ignore my comment above I see the summary_occurrences doesn't get regenerated, it just receives new records.

I am finding it quite difficult to determine if training records are now excluded.
Biren: are you able to go back to the old file please, but keep the new version handy so we can go back to it again.
Thanks

**Comment by BirenRathod** [comment]
@andrewvanbreda The old file is still in place and currently actively working. I haven't made changes yet.

**Comment by andrewvanbreda** [comment]
@BirenRathod 
Ok thanks for letting me know.
Well that would explain the lack of difference, leave it in place for now so I can double-check the data I will be checking befor it is switched over.

**Comment by andrewvanbreda** [comment]
Blimey trying to get my head around this is quite mind-bending.

Firstly @BirenRathod , you just need to know I don't think we need to do the experiment with the Summary Builder I suggested the other day.

I now see in November Gary  made a commit to Github to "Remove training records from summary builder calculations"
https://github.com/Indicia-Team/warehouse/commit/783d22e6da92953b383be54c92b58916f0187fef

So actually my changes from the other day should not be necessary.
Gary's changes actually checks the training flag in fewer places than my code, but as Gary knows the module better than me, this should suffice as I assume he put them in the right places.

I am struggling to setup test data for the Annual Report, but I can only assume @Gary-van-Breda  tested that commit so I don't need to retest it.

A while ago I also changed the download reports on the "Downloads" tab to be training flag sensitive.

This does leave a slightly inconsistent situation where the charts and graphs on the page always exclude training data, but some of the downloads on the Download page show training data depending on the user's own Training setting.

For simplicity I have just added a note to the page to tell the user to always make sure their Training flag is off when viewing the annual report.

With time, a better approach might be to always exclude training data from the downloads, then the message would not be needed, however I would need to explore wheree lse the reports are used before making such a change.

I will send an email to Cristina just to make sure she is happy that training data is not being displayed and also translation thoughts before looking into to getting this issue closed.

Note that this issue is related, as it deals with site-wide issues of data consistency in reports
https://github.com/BiologicalRecordsCentre/ABLE/issues/757

I have raised this equivalent issue for UKBMS to get that site checked
https://github.com/BiologicalRecordsCentre/UKBMS/issues/39
